math/src/polynom/univariate.rs

```diff
@@ -0,0 +1,268 @@
+
+        // compute the zeroth coefficient
+        let c0 = round_claim - self.partial_evaluations[0];
+
+        // compute the evaluations of q
+        let mut q_evals: Vec<E> = vec![E::ZERO; n_minus_1];
+        q_evals.iter_mut().zip(self.partial_evaluations.iter()).enumerate().for_each(
+            |(i, (normalized, evals))| *normalized = (*evals - c0).mul_base(points_inv[i]),
+        );
+
+        // interpolate q
+        let q_coefs = multiply_by_inverse_vandermonde(&q_evals, &points_inv);
+
+        // append c0 to the coefficients of q to get the coefficients of p. The linear term
+        // coefficient is removed as this can be recovered from the other coefficients using
+        // the reduced claim.
+        let mut coefficients = Vec::with_capacity(self.partial_evaluations.len() + 1);
+        coefficients.push(c0);
+        coefficients.extend_from_slice(&q_coefs[1..]);
+
+        UnivariatePolyCoef { coefficients }
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Given a (row) vector `v`, computes the vector-matrix product `v * V^{-1}` where `V` is
+/// the Vandermonde matrix over the points `1, ..., n` where `n` is the length of `v`.
+/// The resulting vector will then be the coefficients of the minimal interpolating polynomial
+/// through the points `(i+1, v[i])` for `i` in `0, ..., n - 1`
+///
+/// The naive way would be to invert the matrix `V` and then compute the vector-matrix product
+/// this will cost `O(n^3)` operations and `O(n^2)` memory. We can also try Gaussian elimination
+/// but this is also worst case `O(n^3)` operations and `O(n^2)` memory.
+/// In the following implementation, we use the fact that the points over which we are interpolating
+/// is a set of equidistant points and thus both the Vandermonde matrix and its inverse can be
+/// described by sparse linear recurrence equations.
+/// More specifically, we use the representation given in [1], where `V^{-1}` is represented as
+/// `U * M` where:
+///
+/// 1. `M` is a lower triangular matrix where its entries are given by M(i, j) = M(i - 1, j) - M(i -
+///    1, j - 1) / (i - 1) with boundary conditions M(i, 1) = 1 and M(i, j) = 0 when j > i.
+///
+/// 2. `U` is an upper triangular (involutory) matrix where its entries are given by U(i, j) = U(i,
+///    j - 1) - U(i - 1, j - 1) with boundary condition U(1, j) = 1 and U(i, j) = 0 when i > j.
+///
+/// Note that the matrix indexing in the formulas above matches the one in the reference and starts
+/// from 1.
+///
+/// The above implies that we can do the vector-matrix multiplication in `O(n^2)` and using only
+/// `O(n)` space.
+///
+/// [1]: https://link.springer.com/article/10.1007/s002110050360
+fn multiply_by_inverse_vandermonde<E: FieldElement>(
+    vector: &[E],
+    nodes_inv: &[E::BaseField],
+) -> Vec<E> {
+    let res = multiply_by_u(vector);
+    multiply_by_m(&res, nodes_inv)
+}
```

Comment:
It is a matrix-vector multiplication so should be parallelizable but since we are talking about small dimensions I am not sure it is worth it. Can investigate it further otherwise.